### DISCUSSION TITLE
Is there a setup that allows me to open the built index.html locally without a server?

### DISCUSSION THREAD

**Discussion by joonaspaakko** [retypeapp/retype#194] [Q&A] [answered]
There is the `url` config, but the asset paths always seem to start with at the very least `/` making it so I have to spin up a server to view the documentation and I'd like an option where I didn't have to do that.

**Comment by fabriciomurta** [comment]
Point [the `url` config](https://retype.com/configuration/project/#url) to `mydomain.com/mysubdir`.

Either way you can always [`retype run`](https://retype.com/guides/cli/#retype-run) or [`retype watch`](https://retype.com/guides/cli/#retype-watch).

The latter is very smart and you can see your rendered page, make edits and, as you save the source file, the page gets updated with latest changes, so you should give it a try.

**Reply by joonaspaakko** [reply]
I can see why it works the way it does, but I was just being hopeful it didn't require a server... I wanted to use retype as internal documentation for a project and I wanted to make reading the documentation as easy as opening `docs/index.html`. Unfortunately, because it requires a server to run at the very least, I have to basically write up another lightweight text documentation to tell other devs how to open the documentation by installing retype and doing retype run, which seems a little cumbersome: "Read the documentation before reading the documentation". But that will have to do.

**Reply by geoffreymcgill** [reply]
This thread is related to https://github.com/retypeapp/retype/discussions/133.

**Reply by geoffreymcgill** [reply]
We have a few ideas that should allow for server-less local browsing of the docs. No time frame on when this will be implemented, although if we can get 100% relative path linking working, then opening the `docs/index.html` will work too. 

This is always the option of publicly hosting your docs too, such as in [GitHub Pages](https://retype.com/guides/github-actions/).

**Comment by DannyBen** [comment]
Modern generated HTML, including Retype, needs javascript, which is not working out of the box for loaded files. Personally, I feel that pursuing an effort to make Retype generated sites work without a server is just going to be a deadweight that the dev team will need to carry around and maintain, with very little benefit or use cases.

Running a public or private static web server (not necessarily the retype command line), is an easy task, and in addition, I think the developers in any group will benefit from having a URL for their documentation, which is always up to date.

There are a couple of static easy-to-install, easy-to-run, webservers that have basic authentication built in, which might be a reasonable solution for setting up a password protected documentation site on a public site.

**Reply by joonaspaakko** [reply]
Obviously I don't know all the things Retype needs for the documentation to run (fully?), but in my test it seemed to work just fine locally when I removed the forward slashes from all resources manually. I don't see why JS would be an issue, if it's just a read only build. The biggest required change (from what I could see) would be to link to the html files directly in order to navigate form one page to the next, because right now it links to folders which would not work without a server. As I don't want to disrupt what Retype already does, I think this could be an extra config or a build Flag. If that is all that is needed, I don't think it would be too much to ask? 

At minimum just removing the forward slash from urls achieves a lot on its own. So if I go from this:

```html
<link href="/resources/css/retype.css?v=1.10.689424686297" rel="stylesheet" data-turbo-track="reload" />
<script type="text/javascript" src="/resources/js/config.js?v=1.10.689424686297" defer data-turbo-track="reload"></script>
<script type="text/javascript" src="/resources/js/retype.js?v=1.10" defer data-turbo-track="reload"></script>
```

to this: 

```html
<link href="resources/css/retype.css?v=1.10.689424686297" rel="stylesheet" data-turbo-track="reload" />
<script type="text/javascript" src="resources/js/config.js?v=1.10.689424686297" defer data-turbo-track="reload"></script>
<script type="text/javascript" src="resources/js/retype.js?v=1.10" defer data-turbo-track="reload"></script>
```

I can already view each html file separately and this is the difference between seeing what's on the left (`with forward slashes`) and what's on the right (`without forward slashes`):

![retype-forward-slash-example](https://user-images.githubusercontent.com/1164476/140501431-e48fd402-c962-469b-abfc-a0037e0de745.png)

Of course, the page links still won't work even without the forward slashes because without a server they need to point to files. And there might be difficulties with certain parts of Retype that I'm not aware of.

-----

I'm not looking to derail the development but I disagree with your arguments against it. I only agree with it if it would complicate things too much but from where I'm standing I don't see how it would be anything but a good addition.

I think running the retype server is about as easy as it gets and there's no need for any other server (unless one is ready at hand, maybe), but it just feels silly having to document how to view the documentation. It's like if someone sends you a Word file and a readme file with instructions to open it by first... let's say enabling a setting in Word before being able to read the file. Or a paper flyer with instructions on how to unfold your paper manual before you can read it. There shouldn't be a need for that.

The hindrance with hosting it is that if you can't use something like Github pages, you'll have to host the documentation separate from the project files and that creates a disconnect between them and forgetting to upload document separately becomes very likely. When instead the documentation could just sit in the project folder where you can edit it if needed using `retype watch` or whatever and anyone could view the built docs without a hassle without having to install retype (not that it's difficult but like I said you need some sort of readme telling others to install retype and run it to view the actual documentation, if it's not hosted).

**Reply by geoffreymcgill** [reply]
We should be able to get local file browsing working. That was part of our original requirements, although we had to drop at some point because we couldn't make it work at the time. We recently ran some simple tests and it does appear to work fine now. 

What's annoying is that I cannot remember exactly why it wasn't working originally. We may rediscover the reason at some point, which would then obviously make local file browsing impossible without a local web server. I vaguely remember an issue with local AJAX requests failing because of browser sandbox limitations, when a web server is not used for hosting. 

@joonaspaakko 

> but it just feels silly having to document how to view the documentation.

The generally accepted practice is that your documentation should be published (hosted) online somewhere. That would also allow the search engines to index your docs. Your source content files (`.md`) would be stored in your source code reposition and your built website (`.hmtl`) are hosting online.

> The hindrance with hosting it is that if you can't use something like Github pages, you'll have to host the documentation separate from the project files and that creates a disconnect between them and forgetting to upload document separately becomes very likely.

The sync'ing of your online documentation should be performed by an automated process, such as GitHub Actions. If you are hosting your website on an external web server and need to FTP the built files, we have an excellent GItHub Action for that. It's not documented yet, but I will get @fabriciomurta to post a tutorial here. The Action will automatically handle the building and synchronization of your documentation project with your FTP web server.

In general, there shouldn't be a need to store the built website files (.html, .css, .js, etc) in your source code repository. Requiring users to build the docs using the Retype App is creating an unnecessary hinderance.

BUT... as I mentioned above, I _think_ we can get relative pathing working, so if you still want to allow local file browsing of the built website, it should work in a future release, maybe `v1.13`.

**Reply by fabriciomurta** [reply]
> I will get @fabriciomurta to post a tutorial here

We do have a drafted action using git-ftp to push FTP it's just not pushed to the [retypeapp org](https://github.com/retypeapp). It's basically a step you add to the [Build Action](https://retype.com/guides/github-actions/) to, instead of pushing to a GitHub Pages repository-branch, do the FTP instead, keeping credentials secure using GitHub's secrets.

So, back to @joonaspaakko quote:

> The hindrance with hosting it is that if you can't use something like Github pages, you'll have to host the documentation separate from the project files and that creates a disconnect between them and forgetting to upload document separately becomes very likely.

So, if you are hosting your project in GitHub (even if in a private repository), you can ensure docs are up to date by using the action; this of course assumes you do use GitHub to host it and, inherently, tracks its documentation changes with _git_ too.

Back up a bit in the post:

> At minimum just removing the forward slash from urls achieves a lot on its own.

If you also fix the links throughout the project, you'll then notice that **search is broken**. I believe that's basically our stone towards the "freedom of paths". There was a lot of discussion among us about this forced absolute pathing and, for one reason or another, the decision was to force the specific path in favor of some dynamics of the built website, like 100% local implementation of the search functionality.

This means, with #133 implemented, this request will likely be fulfilled as well. We have to wait and see about browser limitations on running websites locally though.

I see a very valid point in shipping the documentation with your product as a means to ensure the documentation is up to date with that specific package. Years later you'd be navigating to a website with lots of features not pertaining that version, or in the least (#142) hopefully find the version you are using in a dropdown in the online page.

**Comment by geoffreymcgill** [comment, marked as answer]
Hi @joonaspaakko & @DannyBen,

As of Retype `v2.3.0`, local file browsing is supported, although to enable the functionality requires a couple configs added to your projects `retype.yml` file:

```yml
search:
  preload: true

generator:
  directoryIndex:
    append: true
```

Once added, run the `retype build` command to re-generate the website.

By default, the generated files are added to a `.retype` folder within the root of your project, although that location is configurable using the [`output`](https://retype.com/configuration/project/#output) config.

Once rebuilt, from your local file system should be able to:
1. Double-click on the root `index.html` file to open in a web browser, or
2. Right-click (option-click) on the `index.html` and open the file in a web browser if HTML files are not mapped to automatically open in a web browser, or
3. Use the <kbd>File > Open File...</kbd> menu from within your web browser to select the `index.html` file, or
4. Open via the command line, or
5. Likely a dozen other different ways of doing this same thing.

In general, I would suggest only using the above configs if you want to support local file browsing. If hosting your website on a web server, then you should not require either of the above configs. If you include the two configs, web server hosting will still work perfectly fine, but the experience for the visitors will be better without those two config [if web server hosting].

I apologize for how long this took to implement, but it was not easy to get everything working locally and the new functionality required revisions to almost all corners of Retype.

The upcoming Retype `v2.4.0` release includes a few bug fixes, so you'll want to upgrade to `v2.4.0` once available. 

Hope this helps.

**Reply by joonaspaakko** [reply]
Thanks, it's working as expected. I wasn't really expecting this feature to happen, but I'm glad it did... I already put it to use.
